pkg/kkrtpsi/kkrtpsi.go

```diff
@@ -0,0 +1,135 @@
+package kkrtpsi
+
+import (
+	"encoding/binary"
+	"io"
+	"math"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/optable/match/internal/cuckoo"
+	"github.com/optable/match/internal/hash"
+	"github.com/optable/match/internal/oprf"
+)
+
+const batchSize = 2048
+
+// HashRead reads one hash
+func EncodesRead(r io.Reader, u *[cuckoo.Nhash]uint64) (err error) {
+	err = binary.Read(r, binary.BigEndian, u)
+	return
+}
+
+// HashWrite writes one hash out
+func EncodesWrite(w io.Writer, u [cuckoo.Nhash]uint64) error {
+	return binary.Write(w, binary.BigEndian, u)
+}
+
+type hashEncodingJob struct {
+	batchSize int
+	prc       []oprfEncodeInputs // PseudoRandom Code
+	hashed    [][cuckoo.Nhash]uint64
+	execute   func(job hashEncodingJob)
+}
+
+func makeJob(hasher hash.Hasher, batchSize int, f func(hashEncodingJob)) hashEncodingJob {
+	return hashEncodingJob{
+		batchSize: batchSize,
+		prc:       make([]oprfEncodeInputs, batchSize),
+		execute:   f}
+}
+
+func (bytes oprfEncodeInputs) encodeAndHash(oprfKeys oprf.Key, hasher hash.Hasher) (hashes [cuckoo.Nhash]uint64, err error) {
+	// oprfInput is instantiated at the required size
+	for hIdx, bucketIdx := range bytes.bucketIdx {
+		err = oprfKeys.Encode(bucketIdx, bytes.prcEncoded[hIdx])
+		if err != nil {
+			return hashes, err
+		}
+		hashes[hIdx] = hasher.Hash64(bytes.prcEncoded[hIdx])
+	}
+
+	return hashes, nil
+}
+
+// HashAllParallel reads all identifiers from identifiers
+// and parallel hashes them until identifiers closes
+func EncodeAndHashAllParallel(oprfKeys oprf.Key, hasher hash.Hasher, identifiers <-chan oprfEncodeInputs) <-chan [cuckoo.Nhash]uint64 {
```

Comment:
If we pass the slice of identifers directly, we can really simplify this function. We should remove all of the job management concepts and split the input across the workers as we do in other places in the PR. We should use a workgroup to run both the encoding functions as well as the goroutine writing to the output writer. We could batch the identifiers written to the channel that the writer goroutine reads from to amortize the channel operations.

Reply:
@ratelle Please look at this again. We've used a buffered channel in EncodeAndHashAllParallel but don't have a good intuition by how much to buffer. It seems to perform best with buffering between 1024 and 2048.

Reply:
@ratelle Please look at this again. We've used a buffered channel in EncodeAndHashAllParallel but don't have a good intuition by how much to buffer. It seems to perform best with buffering between 1024 and 2048.

Reply:
I've implemented batching here and used ErrGroup: check [this commit](https://github.com/Optable/match/pull/41/commits/966488ab003641610f71a3e84cc8e500a8c06ce1)